### PR TITLE
#4508 writing tests for workspaceEpics

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4508)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/store/epics/workspace/createWorkspaceEpic.js [hunk 1]
```diff
@@ -54,90 +53,6 @@ const createWorkspace$ = values =>
...
-    map(data => {
-      if (!data?.state || data.state !== SERVICE_STATE.RUNNING) {
-        throw data;
-      }
-      return actions.startWorker.success(data);
-    }),
-    retryWhen(error => error.pipe(delay(duration * 2), take(retry))),
-    catchError(error => of(actions.startWorker.failure(error))),
-  );
-
```
Comment (saivirtue) [inline ohara-manager/client/src/store/epics/workspace/createWorkspaceEpic.js:140]: Using their own logic of services (zookeeper, broker and worker) in each file.

File: ohara-manager/client/src/store/epics/workspace/createWorkspaceEpic.js [hunk 2]
```diff
@@ -166,22 +81,20 @@ export default (action$, state$) =>
...
+          createWorkspace$(workspace),
+          createZookeeper$(zookeeper).pipe(delay(duration)),
+          createBroker$(broker).pipe(delay(duration)),
+          createWorker$(worker).pipe(delay(duration)),
+          startZookeeper$(zookeeper).pipe(delay(duration)),
+          startBroker$(broker).pipe(delay(duration)),
+          startWorker$(worker).pipe(delay(duration)),
+          finalize$(workspaceKey).pipe(delay(duration)),
+        ],
+        asapScheduler,
```
Comment (saivirtue) [inline ohara-manager/client/src/store/epics/workspace/createWorkspaceEpic.js:97]: `asapScheduler` could make sure the "delay" is actually what we want to.

File: ohara-manager/client/src/store/epics/__tests__/workspace/createWorkspaceEpic.test.js [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+    const input = '   ^-a            ';
+    const expected = `--a 2999ms b\
+                      999ms c 1999ms d\
+                      999ms e 1999ms f\
+                      999ms g 1999ms h\
+                      999ms m  499ms n\
+                      999ms o  499ms p\
+                      999ms q  499ms r\
+                      999ms (tuvwxyz)`;
```
Comment (saivirtue) [inline ohara-manager/client/src/store/epics/__tests__/workspace/createWorkspaceEpic.test.js:58]: verbose alias here...but I hope to "group" the same result:
- `c-d` : create zookeeper
- `e-f` : create broker
- `g-h` : create worker
- `m-n` : start zookeeper
- `o-p` : start broker
- `q-r` : start worker
- `t-z` : finalize actions

File: ohara-manager/client/src/store/epics/__tests__/workspace/switchWorkspaceEpic.test.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+    expectSubscriptions(action$.subscriptions).toBe(subs);
+
+    flush();
+
+    expect(spyPush).toHaveBeenCalledTimes(1);
+  });
+});
+
+it('switch to empty workspace should be redirect to head workspace', () => {
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/__tests__/workspace/switchWorkspaceEpic.test.js:87]: should be redirected to the first workspace ❓

File: ohara-manager/client/src/store/epics/broker/startBrokerEpic.js [hunk 5]
```diff
@@ -36,19 +35,16 @@ import * as actions from 'store/actions';
 import * as schema from 'store/schema';
 import { getId } from 'utils/object';
 
-const startBroker$ = params => {
+export const startBroker$ = params => {
   const brokerId = getId(params);
-  return of(
+  return zip(
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/broker/startBrokerEpic.js:40]: @saivirtue could you elaborate a bit why this change 🙏

File: ohara-manager/client/src/store/epics/workspace/createWorkspaceEpic.js (hunk 2 above)
Comment (konekoya) [reply ohara-manager/client/src/store/epics/workspace/createWorkspaceEpic.js:97]: StreamMaster 👍

File: ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js [hunk 6]
```diff
@@ -62,7 +58,9 @@ const startZookeeper$ = params => {
       ),
     ),
   ).pipe(
-    concatAll(),
+    map(([, data]) => normalize(data, schema.zookeeper)),
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js:61]: And so `zip` returns back a flatted stream? Sorry for asking so many dumb questions like, still very new to rxjs XD

File: ohara-manager/client/src/store/epics/__tests__/workspace/switchWorkspaceEpic.test.js (hunk 4 above)
Comment (saivirtue) [reply ohara-manager/client/src/store/epics/__tests__/workspace/switchWorkspaceEpic.test.js:87]: Yep. thanks :)

File: ohara-manager/client/src/store/epics/broker/startBrokerEpic.js (hunk 5 above)
Comment (saivirtue) [reply ohara-manager/client/src/store/epics/broker/startBrokerEpic.js:40]: Using `of` need to "combine" the output result (a.k.a the `concatAll()` part) ; but `zip` will combine the result for us :)

File: ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js (hunk 6 above)
Comment (saivirtue) [reply ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js:61]: Not a problem :)
You are right, zip will return back the flatten result

File: ohara-manager/client/src/store/epics/broker/startBrokerEpic.js (hunk 5 above)
Comment (konekoya) [reply ohara-manager/client/src/store/epics/broker/startBrokerEpic.js:40]: In other words, `zip` creates an observable but `of` creates nested one, right?

File: ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js (hunk 6 above)
Comment (konekoya) [reply ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js:61]: Thanks ❤️

File: ohara-manager/client/src/store/epics/broker/startBrokerEpic.js (hunk 5 above)
Comment (konekoya) [reply ohara-manager/client/src/store/epics/broker/startBrokerEpic.js:40]: Got you answer blow...
